play-json/shared/src/main/scala/play/api/libs/json/JsValue.scala

```diff
@@ -136,7 +137,7 @@ case class JsObject(
    */
   lazy val value: Map[String, JsValue] = underlying match {
     case m: immutable.Map[String, JsValue] => m
-    case m                                 => m.toMap
+    case m                                 => (TreeMap.newBuilder[String, JsValue] ++= m).result()
```

Comment:
On second thought, can we just return `m` directly?

```suggestion
    case m                                 => m
```

The scaladoc says "The value of this JsObject as an _immutable_ map", but the type signature is only `scala.collection.Map`. Anyone requiring a `scala.collection.immutable.Map` would need to call `.toMap` anyway or do an `.asInstanceOf[s.c.i.Map]`.

If we just return `m` directly and change remove "immutable" from the scaladoc, it would do good things for memory footprint and throughput:
```
[info] Benchmark                                          (n)   Mode  Cnt    Score    Error  Units
[info] JsonParsing_01_ParseManyFields.parseObjectValue  10000  thrpt    3  167.452 ± 37.356  ops/s
```

This would shift the responsibility for handling hashCode collisions onto users, but this method isn't suggested usage anyway.

@gmethvin, WDYT?

Reply:
I've created a separate PR for that option: #676.

Reply:
I don't think it's worth the breaking change. Since the method isn't suggested usage I don't think users will benefit much from the change anyway.

Reply:
Looks like `.value` is [getting used](https://github.com/julienrf/play-json-derived-codecs/blob/aa0c78a008e088dbd77768e04a3e633b48f8f24e/library/src/main/scala/julienrf/json/derived/DerivedOWrites.scala#L97) in the wild.

I'll reopen this PR to keep both options on the table.

Reply:
> not worth the breaking change

I'm having difficultly thinking of a scenario where it would be breaking. Method signatures wouldn't change. Anyone doing `.value.asInstanceOf[s.c.i.Map]` should know that's unsound since `JsObject` is non-final and it's already possible to subclass this and make `value` return other representations. Are there other scenarios to be concerned about?

Reply:
yeah, I wouldn't say it's _technically_ a breaking change, but Scala users typically want immutable objects so we should plan for the immutable case.

Reply:
Hmm, okay. How about making `underlying` immutable via a `class ImmutableJMapWrapper[K, V](jmap: java.util.Map[K, V]) extends scala.collection.immutable.Map[K, V]` instead?

Reply:
```scala
/**
  * Wraps a Java String-keyed map as a Scala one with fallback to TreeMap for modifications.
  */
private[json] class ImmutableJMapWrapper[V](underlying: java.util.Map[String, V]) extends scala.collection.immutable.Map[String, V] {

  override def removed(key: String): Map[String, V] = to(TreeMap).removed(key)

  override def updated[V1 >: V](key: String, value: V1): Map[String, V1] = to(TreeMap).updated(key, value)

  override def get(key: String): Option[V] = Option(underlying.get(key))

  override def iterator: Iterator[(String, V)] = new AbstractIterator[(String, V)] {
    private[this] val ui = underlying.entrySet().iterator()

    override def hasNext: Boolean = ui.hasNext

    override def knownSize: Int = if (underlying.isEmpty) 0 else super.knownSize

    override def next(): (String, V) = {
      val e = ui.next()
      (e.getKey, e.getValue)
    }
  }
}
```

Benefits:
- immutable
- replaces JMapWrapper, so same memory footprint.
- no copy on `value`
- third party libraries like play-json-derived-codecs gain access to the same performance that the play-json macros enjoy by using the private `underlying` field

Slightly weird:
- no way to `clone()` the java Map, so modifications need to copy to some other type (AFAIK, only TreeMap is safe against #186)
- `TreeMap` requires an `Ordering`. I just narrowed key to `String` to avoid falling into the next object alignment. `String` is all we need anyway.
- we'd need a separate implementation for scala-2.12-, but it's only 10 lines.

WDYT?

Reply:
How does the performance of `LinkedHashMap` compare to `immutable.TreeMap` in general? I wonder if we should just change the underlying map to `immutable.TreeMap` and avoid using the `LinkedHashMap` hack altogether...

Reply:
Oh, I forgot something potentially important...

IIRC, we use `LinkedHashMap` instead of `HashMap` because some users are expecting a specific ordering of keys. Technically the ordering of keys in JSON isn't significant, but it's something we might want to be careful about breaking.

Reply:
I guess using the `ImmutableJMapWrapper` as the `underlying` map could be a good idea. It feels like an ugly hack but the whole `LinkedHashMap` thing feels that way too.

Reply:
All that said, the "correct" solution here is to fix the bug in Scala, so what we're doing here is hopefully just a "temporary" workaround.

Reply:
> How does the performance of LinkedHashMap compare to immutable.TreeMap in general?

FWIW, `j.u.LinkedHashMap` had 1.3x-1.4x of the memory footprint of `s.c.i.TreeMap` in my testing, but has faster lookup over the common case from the open addressing hash table.

As for `.value`, all of the `s.c.immutable.Map` implementations (except `Map1-4`) destroy order in one way or another on `.toMap`. Our `ImmutableJMapWrapper` would be an exception by preserving order--another benefit.

> All that said, the "correct" solution here is to fix the bug in Scala

That's a good point, although I haven't heard any developments since https://github.com/scala/bug/issues/11203 was marked `won't fix`. I know https://github.com/scala/scala/pull/7633 is available, but that's mutable and doesn't preserve order. 

> It feels like an ugly hack

Well, the [MapAndVector](https://github.com/circe/circe/blob/e6685317bc9e6e65539bb6194900e0e754fbcb7c/modules/core/shared/src/main/scala/io/circe/JsonObject.scala#L384-L387) approach is cleaner, but would be breaking to apply here for anyone [using the case class `unapply`](https://github.com/com-lihaoyi/upickle/blob/106a706287c3f3cfe615cf5c65e692ed81181b87/ujson/play/src/ujson/play/PlayJson.scala#L15) to get at the `underlying` map--which I only just realized was possible.

It's worth mention that I've been experimenting with applying V8's hidden classes to `JsObject` to cut memory footprint. (I love that I can stuff any Map I want into it!) Gains are quite promising for large messages. I'm aware that's probably more crazy than you'd want in the base implementation, but regardless, here are the bones for potential inspiration. It's ordered, as well.

```
class ShapeMap[V](
  values: Array[V],
  sharedShape: Map[String, Int], // array indexes (j.u.HashMap)
  sharedShapes: ConcurrentMap[Seq[String], Map[String, Int]]
) extends scala.collection.immutable.Map[String, V]
```

Anyway, I'm just here to upstream any useful low hanging fruit I'm finding while benching and poking around in the weeds. Do you want me to go ahead with a PR for `ImmutableJMapWrapper`?

Reply:
`ImmutableJMapWrapper` seems like an ok solution.

Honestly it would be really nice if we could just break the ordering and use `TreeMap` as the underlying data structure, but I don't remember exactly what the concerns were there.

Reply:
Happy new year!

I've rewritten the PR to make an immutable wrapper around `java.util.LinkedHashMap`. I coupled directly to LHM rather than all `j.u.Map` since it's easier to support `+` and `++`. Generality seems unnecessary.

I've explored the `Map` space further and ended up using the following implementation (along with specialized Map1-4) in my own project:
```
class SortedArraysMap[V](
  private val sortedKeys: Array[String], // sorted, binary search lookup
  private val vals: Array[V]             // parallel to sortedKeys
) extends scala.collection.immutable.Map[String, V]
```
Everything's mostly `log(n)` (like `TreeMap`), but the footprint is much smaller.
![Map memory footprint by cardinality (1)](https://user-images.githubusercontent.com/663139/148675817-3aaeec90-d6c5-4701-95b9-c0a59a26d2fb.png)

I've relaxed `JsObject.+`, etc. to delegate to the underlying `Map` implementation and avoid forcing `j.u.LHM`. If users don't care about order and want `log(n)` `+`, `++`, etc. instead of `n` with LHM, they have the option of using `TreeMap`, `s.c.i.HashMap`, or something like my `SortedArraysMap`.

Please review when you get a chance. Criticism welcome.